File: rampage/src/main/java/com/rambots4571/rampage/structure/Sequence.java
package com.rambots4571.rampage.structure;

import java.util.Iterator;

public class Sequence<E> implements Iterator<E> {
    private Node<E> first, current, last;
    private int size;

    private static class Node<T> {
        private T element;
        private Node<T> next;

        Node(T element) {
            this.element = element;
        }
    }

    public Sequence() {
        first = null;
        current = null;
        last = null;
        size = 0;
    }

    @Override
    public boolean hasNext() {
        return current.next != null;
    }

    @Override
    public E next() {
        E element = current.element;
        current = current.next;
        return element;
    }

    public void add(E element) {
        Node<E> addend = new Node<>(element);
        if (first == null) {
            first = addend;
            current = addend;
            last = addend;
        } else {
            last.next = addend;
            last = addend;
        }
        size++;
    }

    public int size() {
        return size;
    }
}

File: rampage/src/main/kotlin/com/rambots4571/rampage/structure/Sequence.kt
package com.rambots4571.rampage.structure

class Sequence<E> : Iterator<E> {
    private var first: Node<E>?
    private var current: Node<E>?
    private var last: Node<E>?
    var size: Int
        private set

    private class Node<T>(var element: T) {
        var next: Node<T>? = null
    }

    init {
        first = null
        current = null
        last = null
        size = 0
    }

    override fun hasNext(): Boolean {
        return current?.next != null
    }

    override fun next(): E {
        val e = current?.element
        current = current?.next
        return e!!
    }

    fun add(e: E) {
        val new = Node(e)
        if (first == null) {
            first = new
            current = new
            last = new
        } else {
            last?.next = new
            last = new
        }
        size++
    }
}